pkg/reconciler/eventtype.go

```diff
@@ -0,0 +1,193 @@
+		// Setting the reference to delete the EventType upon uninstalling the source.
+		if err := controllerutil.SetControllerReference(owner, &eventType, r.Scheme); err != nil {
+			return nil, err
+		}
+		eventTypes = append(eventTypes, eventType)
+	}
+	return eventTypes, nil
+}
+
+// makeEventType creates the in-memory representation of the EventType.
+func (r *EventTypeReconciler) makeEventType(spec eventingv1alpha1.EventTypeSpec, namespace string, lbl map[string]string) eventingv1alpha1.EventType {
+	return eventingv1alpha1.EventType{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("%s-", ToDNS1123Subdomain(spec.Type)),
+			Labels:       lbl,
+			Namespace:    namespace,
+		},
+		Spec: spec,
+	}
+}
+
+// getSyncEventTypes computes the EventTypes that need to be created and/or deleted based on the difference between
+// 'expected' and 'current'. It does so using all the EventType.Spec fields but Description.
+func (r *EventTypeReconciler) getEventTypesToSync(current []eventingv1alpha1.EventType, expected []eventingv1alpha1.EventType) *syncEventTypes {
+	eventTypesToSync := &syncEventTypes{
+		ToCreate: make([]eventingv1alpha1.EventType, 0),
+		ToDelete: make([]eventingv1alpha1.EventType, 0),
+	}
+	keyFunc := func(eventType *eventingv1alpha1.EventType) string {
+		return fmt.Sprintf("%s_%s_%s_%s", eventType.Spec.Type, eventType.Spec.Source, eventType.Spec.Schema, eventType.Spec.Broker)
+	}
+
+	currentMap := groupByKey(current, keyFunc)
+	for _, e := range expected {
+		if et, ok := currentMap[keyFunc(&e)]; !ok {
+			// If it's not in the currentMap, we need to create it.
+			eventTypesToSync.ToCreate = append(eventTypesToSync.ToCreate, e)
+		} else {
+			// If it's in the currentMap and there are more than one,
+			// then we just keep the first one, and delete the others.
+			if len(et) > 1 {
+				for i := 1; i < len(et); i++ {
+					eventTypesToSync.ToDelete = append(eventTypesToSync.ToDelete, et[i])
+				}
+			}
+		}
+	}
+	return eventTypesToSync
+}
+
+// groupByKey groups 'eventTypes' by key, where the key is given by 'keyFunc'.
+func groupByKey(eventTypes []eventingv1alpha1.EventType, keyFunc func(*eventingv1alpha1.EventType) string) map[string][]eventingv1alpha1.EventType {
+	eventTypesAsMap := make(map[string][]eventingv1alpha1.EventType, 0)
+	for _, eventType := range eventTypes {
+		key := keyFunc(&eventType)
+		eventTypesAsMap[key] = append(eventTypesAsMap[key], eventType)
+	}
+	return eventTypesAsMap
+}
+
+// Converts 'name' to a valid DNS1123 subdomain, required for object names in K8s.
+func ToDNS1123Subdomain(name string) string {
+	// If it is not a valid DNS1123 subdomain, make it a valid one.
+	if msgs := validation.IsDNS1123Subdomain(name); len(msgs) != 0 {
+		// If the length exceeds the max, cut it and leave some room for a potential generated UUID.
+		if len(name) > validation.DNS1123SubdomainMaxLength {
+			name = name[:validation.DNS1123SubdomainMaxLength-10]
```

Comment:
Should this `10` be a constant? Maybe easier to document.

Reply:
Done! thanks!